src/ReinforcementLearningZoo/src/algorithms/policy_gradient/vmpo.jl

```diff
@@ -0,0 +1,235 @@
+    clamp.(a, -m, m) |> send_to_host |> first
+end
+
+function (policy::VMPOPolicy{<:ActorCritic{<:GaussianNetwork},Normal})(
+    state::AbstractArray,
+    action::AbstractArray,
+)
+    μ, logσ = policy.approximator.actor(policy.rng, state)
+    action = atanh.(action)
+    μ, logσ, normlogpdf(μ, exp.(logσ), reshape(action, size(μ)...))
+end
+
+## update policy
+
+function RLBase.update!(p::VMPOPolicy, t::VMPOTrajectory, env::AbstractEnv, ::PreActStage)
+    # in the first update, only state & action are inserted into trajectory
+    length(t) == 0 && return
+
+    p.update_step += 1
+    if p.update_step % p.update_freq == 0
+        _update!(p, t)
+    end
+end
+
+function _update!(p::VMPOPolicy, t::VMPOTrajectory)
+    AC = p.approximator
+    D = device(AC)
+    s = send_to_device(D, @view t[:state][:, 1:end-1])  # drop the last extra state
+    a = send_to_device(D, @view t[:action][1:end-1])  # drop the last extra action
+    is_discrete = isa(p, VMPOPolicy{<:ActorCritic,Categorical})
+    π_old = is_discrete ? AC.actor(s) |> softmax : nothing
+    μ_old, σ_old = is_discrete ? (nothing, nothing) : begin
+        μ, logσ = AC.actor(s)
+        μ, exp.(logσ)
+    end
+
+    # normalise rewards based on each trajectory
+    inds = vcat(0, findall(t[:terminal]), length(t[:terminal]))
+    rewards =
+        map(
+            Flux.normalise,
+            view.(Ref(t[:reward]), (:).(inds[1:end-1] .+ 1, inds[2:end])),
+        ) |>
+        Iterators.flatten |>
+        collect
+
+    # calculate discounted rewards
+    rewards = discount_rewards(rewards, p.γ, terminal = t[:terminal])
+    rewards = send_to_device(D, rewards)
+
+    # calculate advantages and sample top half of it
+    advantages = rewards .- AC.critic(s)[1, :]
+    top_k_idx = sortperm(advantages |> send_to_host, rev = true)[1:length(advantages)÷2]
+    top_k_advs = advantages[top_k_idx]
+
+    for epoch in 1:p.n_epochs
+        ps = Flux.params(AC, p.η, p.α)
+        gs = gradient(ps) do
+            η, α, αμ, ασ = p.η[1], p.α...
+            π, logπ = is_discrete ? p(s, a) : (nothing, nothing)
+            μ, σ, logπ =
+                is_discrete ? (nothing, nothing, logπ) : begin
+                    μ, logσ, logπ = p(s, a)
+                    μ, exp.(logσ), logπ
+                end
+
+            # critic loss
+            v = AC.critic(s)[1, :]
+            Lᵥ = Flux.Losses.mse(v, rewards) / 2
+
+            # actor loss
+            ψ = softmax(top_k_advs / dropgrad(η))
+            Lπ = -ψ ⋅ logπ[top_k_idx]
+
+            # loss for temperature η
+            Lη = η * p.ϵ_η + η * log(mean(exp.(top_k_advs / η)))
+
+            # loss for Lagrange multiplier α
+            Lα = if is_discrete
+                π_pairs = ignore() do
+                    zip(eachcol(π_old), eachcol(π))
+                end
+                KL = [Flux.kldivergence(i...) for i in π_pairs]
+                mean(α * (p.ϵ_α .- dropgrad(KL)) + dropgrad(α) * KL)
```

Comment:
The usage of `Flux.kldivergence` is kind of weird to me. I think it supports 2d array, right? Here it seems you only need to set the `agg=identity`.

Reply:
That's valid, I've made the change. Thanks.